acme/acme.go

```diff
@@ -328,29 +355,44 @@ func (a *ACME) CreateLocalConfig(tlsConfig *tls.Config, checkOnDemandDomain func
 func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
 	domain := types.CanonicalDomain(clientHello.ServerName)
 	account := a.store.Get().(*Account)
-	//use regex to test for wildcard certs that might have been added into TLSConfig
-	for k := range a.TLSConfig.NameToCertificate {
-		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
-		match, _ := regexp.MatchString(selector, domain)
-		if match {
-			return a.TLSConfig.NameToCertificate[k], nil
-		}
-	}
+
+	// Check for existing challenge cert
+	log.Debugf("Checking ACME challenge for %s", domain)
 	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
-		log.Debugf("ACME got challenge %s", domain)
+		log.Debugf("Returning ACME challenge cert for %s", domain)
 		return challengeCert, nil
 	}
+
+	// Check for existing ACME domain cert
+	log.Debugf("Checking ACME domain cert for %v", domain)
 	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
-		log.Debugf("ACME got domain cert %s", domain)
+		log.Debugf("Returning ACME domain cert for %s", domain)
 		return domainCert.tlsCert, nil
 	}
+
 	if a.OnDemand {
 		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
 			return nil, nil
 		}
-		return a.loadCertificateOnDemand(clientHello)
+		cert, err := a.loadCertificateOnDemand(clientHello)
+		if cert != nil && err == nil {
+			return cert, nil
+		}
+		log.Errorf("On demand certicate retrival failed for %v due to %v", domain, err)
 	}
-	log.Debugf("ACME got nothing %s", domain)
+
+	// Check if we have a wildcard cert into TLSConfig that could be used
+	log.Debugf("Checking wildcard certificate matching for %v", domain)
+	for k := range a.TLSConfig.NameToCertificate {
```

Comment:
I'm wondering if by doing this it will conflict with the `IgnoreFilter` used to allow or not LE certificate requests.

Example:

I have `*.domain.tld` which is self signed. I want `fourme.domain.tld` to use LE.

If I set the `IgnoreFilter` to `fourme.domain.tld`, given your statement:
- it will not use the self-signed cert.
- it will not request a LE as per `LoadCertificateForDomains` func.

The only way to make it works is to define the domain _inside_ the toml configuration (which is not filtered AFAIK).

The main usage for `IgnoreFilter` was meant to filter out domains for which I don't want any LE generation. 

Am I missing something?